File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/arch_units/cpp/au_071_c_div_i32_f64.cpp
// arch-unit 71  --  c  `a / b`  lhs=int32_t rhs=double
// symbol op_214   outcome LIFTED
//
// the arch-unit, arch-opcode by arch-opcode:
//   fcvt.d.w fa5, a0                   float    emulation:i32_to_f64_rm0
//   fdiv.d fa0, fa5, fa0, dyn          float    emulation:f64_div_rm0
//   c.jr ra                            integer  return
//
// answer: f64, 64 bits.  parameters are operand bit patterns.
#include "arch_units.hpp"

namespace archunits {

uint64_t au_071_c_div_i32_f64(uint64_t p0, uint64_t p1)
{
    /* a0: operand `a` (int32_t) sign-extended to XLEN */
    const uint64_t v1 = (uint64_t)(int64_t)(int32_t)(uint32_t)(p0);
    /* fa0: operand `b` (double) arrives in fa0 */
    const uint64_t v2 = p1;
    const uint64_t v3 = sfemul::i32_to_f64_rm0((uint32_t)(v1));
    const uint64_t v4 = sfemul::f64_div_rm0(v3, v2);
    return v4;
}

}  // namespace archunits

File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/arch_units/js/au_071_c_div_i32_f64.js
// arch-unit 71  --  c  `a / b`  lhs=int32_t rhs=double
// symbol op_214   outcome LIFTED
//
// the arch-unit, arch-opcode by arch-opcode:
//   fcvt.d.w fa5, a0                   float    emulation:i32_to_f64_rm0
//   fdiv.d fa0, fa5, fa0, dyn          float    emulation:f64_div_rm0
//   c.jr ra                            integer  return
//
// answer: f64, 64 bits.  parameters are operand bit patterns.
'use strict';
const AF = require('./au_float.js');

function au_071_c_div_i32_f64(p0, p1) {
  // a0: operand `a` (int32_t) sign-extended to XLEN
  const v1 = (((((p0) & 0xffffffffn) ^ 0x80000000n) - 0x80000000n) & 0xffffffffffffffffn);
  // fa0: operand `b` (double) arrives in fa0
  const v2 = p1;
  const v3 = AF.i32_to_f64_rm0(((v1) & 0xffffffffn));
  const v4 = AF.f64_div_rm0(v3, v2);
  return v4;
}

module.exports = { au_071_c_div_i32_f64 };
